transport/http2_client.go

```diff
@@ -922,26 +921,51 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame, isFirst bool) {
 	if f.IsAck() {
 		return
 	}
-	var ss []http2.Setting
+	var rs []http2.Setting
+	var ps []http2.Setting
 	isMaxConcurrentStreamsMissing := true
 	f.ForeachSetting(func(s http2.Setting) error {
 		if s.ID == http2.SettingMaxConcurrentStreams {
 			isMaxConcurrentStreamsMissing = false
 		}
-		ss = append(ss, s)
+		if t.isRestrictive(s) {
+			rs = append(rs, s)
+		} else {
+			ps = append(ps, s)
+		}
 		return nil
 	})
 	if isFirst && isMaxConcurrentStreamsMissing {
 		// This means server is imposing no limits on
 		// maximum number of concurrent streams initiated by client.
 		// So we must remove our self-imposed limit.
-		ss = append(ss, http2.Setting{
+		ps = append(ps, http2.Setting{
 			ID:  http2.SettingMaxConcurrentStreams,
 			Val: math.MaxUint32,
 		})
 	}
-	// The settings will be applied once the ack is sent.
-	t.controlBuf.put(&settings{ack: true, ss: ss})
+	t.applySettings(rs)
+	t.controlBuf.put(&settingsAck{})
+	t.applySettings(ps)
+}
+
+func (t *http2Client) isRestrictive(s http2.Setting) bool {
+	var res bool
+	switch s.ID {
+	case http2.SettingMaxConcurrentStreams:
+		t.mu.Lock()
+		if int(s.Val) < t.maxStreams {
+			res = true
+		}
+		t.mu.Unlock()
+	case http2.SettingInitialWindowSize:
+		t.mu.Lock()
+		if s.Val < t.streamSendQuota {
```

Comment:
Shouldn't we be comparing this to `t.initialWindowSize`?

Also, do we have to take a lock for this field or `t.maxStreams`?  I was thinking the only thing that should be changing these fields is the handling of a settings frame (so, this code here)...

Reply:
`t.initialWindowSize` controls our inbound window size while `t.streamSendQuota` controls our outbound window size. So, when we get a settings frame to update our initial window size( the outbound one) we update this variable.

You're right about `t.maxStreams`; updating it never required a lock. I copied the original code and made the same mistake!